src/PathMappingRepository.php

```diff
@@ -0,0 +1,547 @@
+     */
+    public function remove($query, $language = 'glob')
+    {
+        throw new BadMethodCallException(sprintf('%s\(\) is not supported in %s', __METHOD__, __CLASS__));
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function clear()
+    {
+        // Subtract root
+        $removed = $this->countStore() - 1;
+
+        $this->store->clear();
+        $this->createRoot();
+
+        return $removed;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function listChildren($path)
+    {
+        return new ArrayResourceCollection($this->getChildren($this->get($path)));
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function hasChildren($path)
+    {
+        return $this->getChildren($this->get($path))->count() !== 0;
+    }
+
+
+    /**
+     * Find a resource by its path.
+     *
+     * @param string $path The path to resolve.
+     *
+     * @return \Puli\Repository\Api\Resource\Resource|null  The resource or null if
+     *                                                      the resource is not found.
+     */
+    private function resolveResource($path)
+    {
+        /*
+         * If the path exists in the store, return it directly
+         */
+        if ($this->store->exists($path)) {
+            $resolved = $this->createResource($this->store->get($path));
+            $resolved->attachTo($this, $path);
+
+            return $resolved;
+        }
+
+        /*
+         * Otherwise, we need to "resolve" it in two steps:
+         *      1.  find the resources from the store that are potential parent
+         *          of the path (we find them using regular expressions)
+         *      2.  for each of these potential parent, we try to find a real
+         *          file or directory on the filesystem and if we do find one,
+         *          we stop
+         */
+        $storePaths = array_reverse($this->store->keys());
+
+        // Create regex
+        $regExpressions = array();
+
+        foreach ($storePaths as $storePath) {
+            $prefix = rtrim($storePath, '/') . '/';
+            $regExpressions[] = '~^' . preg_quote($prefix, '~') . '~';
+        }
+
+        // Find potential parents
+        $resolved = null;
+
+        foreach ($regExpressions as $key => $regExpression) {
+            if (preg_match($regExpression, $path)) {
+                // The current resource is a potential parent, let's check if it's
+                // a real one by checking the filesystem
+                $filesystemRoot = rtrim($this->store->get($storePaths[$key]), '/') . '/';
+                $filesystemPath = preg_replace($regExpression, $filesystemRoot, $path);
+
+                $resource = $this->createResource($filesystemPath);
+
+                // The resource is resolved
+                if ($resource instanceof FilesystemResource) {
+                    $resolved = $resource;
+                    $resolved->attachTo($this, $path);
+                    break;
```

Comment:
You can directly return here.

What happens when accessing a virtual resource? It should be possible to access them too, no?